### PR TITLE
fix(eventhubs): stop delivery and release ownership on shutdown

### LINKED ISSUE #5096: [Event Hubs] EventProcessor::shutdown() does not stop event delivery or release ownership
## Summary

`EventProcessor::shutdown()` sets a flag and nothing else. It does not close the receivers in the consumers map and it does not release the ownership records, so a partition client that a caller already holds keeps delivering events after `run()` returns.

## Motivation

A caller reads `shutdown()` as "stop processing". It does not stop processing. The whole body sets `is_running = false`, and `run()` reads that flag only after the `update_interval` sleep, so the loop can take a full interval to notice. Every receiver the processor opened stays attached, and every ownership record stays live in the store, so another instance cannot take those partitions until the records expire. `EventProcessor::close(self)` closes receivers, but it consumes the value out of an `Arc`, which a caller holding a partition client cannot do, and it skips any client the application still holds. `ProcessorConsumersMap::revoke_partition_clients` also closes receivers, but only on the load balancer steal path. Neither path runs when the caller asks the processor to stop. A further obstacle: `run()` builds the consumers map as a local value and stores it nowhere, so `shutdown(&self)` has no path to the receivers at all, and the fix needs that map to live on the processor. The .NET processor stops delivery when it stops.

## Proposal

- Close the receivers in the consumers map as part of `shutdown()`, so event delivery stops when the caller asks for it.
- Release the ownership records the instance holds, so another instance can claim those partitions without waiting for the expiry.
- Document the latency that the `update_interval` sleep adds before `run()` observes the flag.
- If `shutdown()` must stay a flag only, rename it or document plainly that it stops new work and not delivery.

### Reference

The Go SDK already does both steps, and the order matters: it closes every issued partition client first, then relinquishes ownership by writing an empty owner id to the checkpoint store. See `Processor.Run` in `sdk/messaging/azeventhubs/processor.go` in azure-sdk-for-go, which ranges over the consumers map calling `client.Close(ctx)`, then sets `OwnerID = relinquishedOwnershipID` (the empty string) on every current ownership and calls `ClaimOwnership`.

The .NET reference documentation for `EventProcessorClient.StopProcessingAsync` states that the processor updates the ownership of its partitions and cleans up the network resources, and that stopping performs network I/O and may wait for active partition reads to complete. It does not say whether the updated record is immediately claimable by a peer, so treat immediate re-claimability as a design choice for this crate rather than as .NET parity.

### PR DESCRIPTION
**Pull request by j7nw4r**
## Summary

`EventProcessor::shutdown` now stops the event delivery and releases the ownership records that the instance holds. It set an internal flag and did nothing else before, so a partition client that the application already held kept delivering events after `run` returned.

## Motivation

`shutdown` set `is_running = false` and stopped there. `run` built its consumers map as a local value and stored it nowhere, so `shutdown` could reach no receiver at all. Every receiver stayed attached, and every ownership record stayed live, so a second instance could not claim those partitions until the records expired. `close` closed receivers, but it consumes the processor out of its `Arc`, which a caller that holds a partition client cannot do.

`shutdown` now guarantees two things. It stops the delivery on every partition client that the processor issued, including one the application still holds, and the `stream_events` stream of such a client resolves with `ConsumerDisconnected`. It then releases the ownership records whose owner is this instance, so another instance can claim those partitions without a wait for the expiration. A failed release logs at the warn level and does not fail the call.

`close` differs from `shutdown` in what it adds, not in what it stops. `close` runs the same stop path, and it also consumes the processor, drains the queued partition clients, and closes the consumer client. `close` is therefore a superset of `shutdown`.

## Changes

- `EventProcessor` keeps its consumers map on `self`, and `run` uses that map in place of a local one.
- Added `ProcessorConsumersMap::close_all_receivers`, which closes each receiver and removes no map entry, so a partition client that the application retains keeps its place.
- Added a private `stop` that `shutdown` and `close` share. It drops the `is_running` guard before it awaits, because a `std::sync::MutexGuard` is not `Send` and holding it across an await deadlocks.
- The ownership release keeps only the records whose owner id matches this instance, and it keeps the ETag that the store returned, because `claim_ownership` rejects a stale ETag.
- Documented that `run` reads the shutdown flag only after the `update_interval` sleep, so `run` can take up to one full `update_interval` to return, while the delivery stops as soon as `shutdown` returns.
- Added a CHANGELOG entry under the unreleased version.

## Test plan

- Five offline tests in the `mod tests` of `processor.rs` were written first and proved red against the unchanged source: `shutdown_closes_receivers_of_issued_partition_clients`, `shutdown_releases_only_this_instances_ownerships`, `shutdown_continues_when_the_ownership_release_fails`, `shutdown_twice_succeeds_and_keeps_ownership_released`, and `close_runs_the_shutdown_stop_path`.
- Each delivery test installs an offline `EventReceiver` before it asserts, because a partition client with an empty receiver returns a canned error stream that pins nothing. An unclosed offline receiver yields `AmqpError`, and only the closed path yields `ConsumerDisconnected(None)`.
- `shutdown_future_is_send` guards the deadlock. It becomes a compile error, not a failed assertion, if a later change holds the `is_running` guard across an await.
- `close_continues_past_a_retained_partition_client` still passes, which holds the contract that a retained partition client stays in the consumers map.
- `CARGO_BUILD_JOBS=1 cargo test --package azure_messaging_eventhubs --lib -- --test-threads=1` reports 150 passed, 0 failed, 14 ignored.
- `CARGO_BUILD_JOBS=1 RUSTFLAGS=-Dwarnings cargo test --no-run --package azure_messaging_eventhubs` exits 0. `cargo fmt --check` and `cargo clippy --all-targets` are clean.
- Assertions were added inside the existing live test `receive_events_from_processor`, which asserts that a retained partition client stops within 30 seconds of `shutdown`. No live test ran in this session, so that assertion waits for a live pass.

Closes #5096

### CODE REVIEW

File: sdk/eventhubs/azure_messaging_eventhubs/src/event_processor/processor.rs [hunk 1]
```diff
@@ -299,17 +329,89 @@ impl EventProcessor {
...
+        {
+            let mut is_running = self.is_running.lock().map_err(|_| {
+                EventHubsError::with_message("Failed to acquire lock on is_running for shutdown")
+            })?;
+            *is_running = false;
+        }
 
-        *is_running = false;
+        self.consumers.close_all_receivers().await?;
+        self.release_ownerships().await;
```
[the remark this answers is not in the document]
Comment (j7nw4r) [reply sdk/eventhubs/azure_messaging_eventhubs/src/event_processor/processor.rs:371]: You are right, and I have fixed it. `stop` takes no lock that `dispatch` also takes, and `dispatch` never read `is_running`, so the two really do interleave. The dispatch task holds the load balancer lock for its whole body (processor.rs:451), and `stop` never asks for that lock. One half is worse than you describe. Even when `close_all_receivers` sees the client, `request_close_receiver` reads an empty `OnceLock` and closes nothing, because `add_partition_client` puts the client in the map before it opens the receiver.

I chose a different fix from the one you suggest. `dispatch` can park on the bounded partition client queue while it holds the load balancer lock. A `stop` that waits for that lock would hang the shutdown, and `close` drains that queue only after `stop` returns. So the dispatch now cleans up after itself. It re-reads `is_running` after the load balancer claim, and again after the receiver opens. The mutex around that flag gives a total order. Either the dispatch sees the stop and undoes its own work, or the stop's list and its map snapshot both run after the dispatch published that work and pick it up. One side always cleans up.

Two unit tests pin the behavior. I proved both by deleting one check at a time and watching only the matching test go red.

**Comment by sagar0207** [comment]
One lifecycle edge case looks worth covering before merge: reusing the same `EventProcessor` after `shutdown()`.

Before this change, every `run()` created a fresh `ProcessorConsumersMap`. The map now lives on `EventProcessor`, and `close_all_receivers()` deliberately retains its entries. `add_partition_client()` treats any existing key as a live client, even when the stored `Weak<PartitionClient>` no longer upgrades, while a retained client's `EventReceiver` has been permanently closed and cannot be replaced because it is stored in a `OnceLock`.

As a result, a subsequent `run()` can reclaim ownership but skip creating usable partition clients. Is `shutdown()` intended to be terminal for an `EventProcessor` instance? If so, could that be documented and enforced by rejecting another `run()`? If restart is intended to work, the map needs generation/identity-aware cleanup (so an old client cannot later remove a replacement), plus a `shutdown -> run again` test.
